packages/contracts/contracts/zkopru/controllers/Migratable.sol

```diff
@@ -1,84 +1,127 @@
+        Migratable(source).transfer(migrationRoot, migration, index, siblings);
+        // 3. fetch new balances
+        uint256 newEthBal = address(this).balance;
+        uint256 newTokenBal;
+        if (token != address(0)) {
+            newTokenBal = IERC20(token).balanceOf(address(this));
+        }
+        // 4.ccheck asset transfer validity
+        require(
+            prevEthBal.add(migration.asset.eth) == newEthBal,
+            "Invalid ETH transfer"
+        );
+        require(
+            prevTokenBal.add(migration.asset.amount) == newTokenBal,
+            "Invalid Token transfer"
+        );
+        //  5. Check Mass Deposit Leaves
+        bytes32 merged;
+        for (uint256 i = 0; i < leaves.length; i++) {
+            merged = keccak256(abi.encodePacked(merged, leaves[i]));
         }
+        require(
+            merged == migration.depositForDest.merged,
+            "Data Availability problem."
+        );
+        // 6. Record the carried mass deposit as committed
+        bytes32 migrationHash = migration.hash();
+        bytes32 depositHash = massDeposit.hash();
+        Storage.chain.committedDeposits[depositHash] += 1;
+        emit NewMassMigration(source, migrationRoot, migrationHash);
     }
 
-    function acceptMigration(
-        bytes32 checksum,
-        bytes32 merged,
-        uint256 fee
-    ) external virtual {
+    function transfer(
+        bytes32 migrationRoot,
+        MassMigration memory migration,
+        uint256 index,
+        bytes32[] memory siblings
+    ) external onlyDestination(migration.destination) {
+        // 1. verify inclusion proof.
+        uint256 _migrationRoot = uint256(migrationRoot);
+        uint256 _leaf = uint256(migration.hash());
+        uint256[] memory _siblings;
+        assembly {
+            _siblings := siblings
+        }
+        require(
+            Storage.chain.migrationRoots[migrationRoot],
+            "Migration root does not exist."
+        );
         require(
-            Storage.allowedMigrants[msg.sender],
-            "Not an allowed departure"
+            getHasher().merkleProof(_migrationRoot, _leaf, index, _siblings),
+            "Failed to verify its inclusion."
         );
-        Storage.chain.committedDeposits[MassDeposit(merged, fee).hash()] += 1;
-        emit NewMassMigration(checksum, msg.sender, merged, fee);
+        // 2. check it's not transferred yet.
+        require(
+            !Storage.chain.transferredMigrations[migrationRoot][bytes32(_leaf)],
+            "Already transferred."
+        );
+        // 3. record it as transferred.
+        Storage.chain.transferredMigrations[migrationRoot][
+            bytes32(_leaf)
+        ] = true;
+        // 4. transfer assets.
+        // send ETH
+        payable(migration.destination).transfer(migration.asset.eth);
+        // send ERC20
+        if (migration.asset.token != address(0)) {
+            IERC20(migration.asset.token).transfer(
```

Comment:
This call should be wrapped in a require statement, ERC20 transfers don't revert on failure, they return either true or false.

Reply:
This is using `SafeERC20`, it should use the `safeTransfer` function or define what happens if the `transfer` function returns false. Tokens like `AMPL` come to mind with strange supply dynamics.

Reply:
NVM, was looking at an old revision.

Reply:
Thanks resolved by https://github.com/zkopru-network/zkopru/pull/206/commits/4a125758c92afe09626df3548266b8a81df132aa